specification/dartLangSpec.tex

```diff
@@ -826,7 +826,8 @@ \section{Variables}
 
 <declaredIdentifier> ::= <metadata> \COVARIANT{}? <finalConstVarOrType> <identifier>
 
-<finalConstVarOrType> ::= \FINAL{} <type>?
+<finalConstVarOrType> ::= \LATE{} \FINAL{}? <type>?
```

Comment:
For processing efficiency, we have exactly one modifier sequence. `external` must appear before `static` or `covariant` which much appear before `final`, etc.

While we could add this as the first modifier grouping that can appear in either order, I'd rather not open that door. Please pick one modifier order and allow the other to be an error/warning.

Reply:
In other places we're pretty opinionated about ordering (e.g. `static final`) and someone (I think @bwilkerson or @stereotype441) was claiming that the Java community started enforcing this as a convention at some point just because the inconsistency was painful.  @munificent thoughts on this?

Reply:
That was me.

Reply:
I'm reverting the changes I introduced into Dart.g based on Lasse's proposal to swap `final` and `late` and deleted the 'Suggested change' that I used to have in the previous comment.  Similar changes were made in other comments below.

Reply:
I can see a number of patterns emerge for myself after trying to convert code to NNBD:
* Local variables differ. I usually write `late final` without an intializer (late assignment), and `final late` with an initializer (lazy initialization).
* For class-level variables, I always write `final late` (because they are late assignments).
* No top-level variables yet. 

All in all, if we only allow one ordering, I think I'd prefer `final late` to `late final`, after having actually written code with it.

Reply:
Do you want to allow `late var x`, `var late x`?  What about `int late x` and `late int x`?  I'm a little confused about how to think about this.

Reply:
Never had to think about `var`, and I can see how that's confusing.

I use `late final type name` or `final late type name`. If I omit the type (which only makes sense for a declaration with an initializer), it will be `final late name = ...`.

Since `var` is different from `final` in that you cannot have a type with `var`, the options are only:
* `var late x;` and
* `late var x;`.
The latter *reads* better because you actually read it as "late variable x", which works with natural language, where "variable late x" does not. That is, "var" occurs as a noun when you read it.
That's unlike "final" which is read as an adjective, and therefore can be swapped with the adjective "late".

Modifiers before type, so definitely `late int x;`, not `int late x`, and I treat `var` as a placeholder for a type, not a modifier, so probably `late var x` too.
(I would be perfectly happy if we allowed `final var x`, and treated `final x` as a shorthand for that, then `var` really *is* a placeholder for a type).

So, what I want is:
 * `late int x`
 * `late var x`
 * `final int x`
 * `final x`
 * `final late x`
 * `final late int x`
 * `late final x`
 * `late final int x`

If I can't get both orderings, then I'd prefer `final late (int) x` to `late final (int) x`. 
There is no technical issue with allowing both, so it would be because someone wants there to be only one way to do things. You can enforce that with lints if they really want it, no need to restrict the language for everybody else (or lock down the style guide before people have had a chance to actually use the feature).

Reply:
> ... no need to restrict the language for everybody else ...

I was not particularly arguing to restrict the language (not that I haven't on other occasions :-). I merely pointed out two things of possible interest.

First, that Dart imposes a single order for all other sequences of modifiers and that we might want to be consistent. If we think that it's worth allowing the modifiers `late` and `final` to appear in either order, then perhaps we should allow other modifiers to appear in arbitrary order as well.

Second, that Java started out allowing modifiers to appear in any order, and the community strongly preferred having a single order for them. This suggests that having a single order for modifiers might improve readability rather than reduce it, and hence it might make sense to specify a single order for these modifiers just as we've done elsewhere in the grammar.

Reply:
We resolved this offline in favor of `late final` being the canonical ordering.